src/mlpack/tests/q_learning_test.cpp

```diff
@@ -0,0 +1,126 @@
+#include <mlpack/methods/ann/ffn.hpp>
+#include <mlpack/methods/ann/init_rules/gaussian_init.hpp>
+#include <mlpack/methods/ann/layer/layer.hpp>
+#include <mlpack/methods/reinforcement_learning/q_learning.hpp>
+#include <mlpack/methods/reinforcement_learning/environment/mountain_car.hpp>
+#include <mlpack/methods/reinforcement_learning/environment/cart_pole.hpp>
+#include <mlpack/methods/reinforcement_learning/policy/greedy_policy.hpp>
+
+#include <boost/test/unit_test.hpp>
+#include "test_tools.hpp"
+
+using namespace mlpack;
+using namespace mlpack::ann;
+using namespace mlpack::optimization;
+using namespace mlpack::rl;
+
+BOOST_AUTO_TEST_SUITE(QLearningTest);
+
+//! Test DQN in Cart Pole task.
+BOOST_AUTO_TEST_CASE(CartPoleWithDQN)
+{
+  // Set up the network.
+  FFN<MeanSquaredError<>, GaussianInitialization> model;
+  model.Add<Linear<>>(4, 128);
+  model.Add<ReLULayer<>>();
+  model.Add<Linear<>>(128, 128);
+  model.Add<ReLULayer<>>();
+  model.Add<Linear<>>(128, 2);
+
+  // Set up the optimizer.
+  StandardSGD<decltype(model)> opt(model, 0.0001, 2);
+
+  // Set up the policy and replay method.
+  GreedyPolicy<CartPole> policy(1.0, 1000, 0.1);
+  RandomReplay<CartPole> replayMethod(10, 10000);
+
+  // Set up DQN agent.
+  QLearning<CartPole, decltype(model), decltype(opt), decltype(policy)>
+          agent(model, opt, 0.9, policy, replayMethod, 100, 100, false, 200);
+
+  arma::running_stat<double> averageReturn;
+  while (true)
+  {
+    double episodeReturn = agent.Episode();
+    averageReturn(episodeReturn);
+    /**
+     * Reaching running average return 35 is enough to show it works.
+     * For the speed of the test case, I didn't set high criterion.
+     */
+    Log::Debug << "Average return: " << averageReturn.mean() << " Episode return: "
+        << episodeReturn << std::endl;
+    if (averageReturn.mean() > 35)
+    {
+      agent.Deterministic() = true;
+      arma::running_stat<double> testReturn;
+      for (size_t i = 0; i < 10; ++i)
+        testReturn(agent.Episode());
+      Log::Debug << "Average return in deterministic test: " << testReturn.mean()
+          << std::endl;
+      break;
+    }
+  }
+}
+
+//! Test Double DQN in Cart Pole task.
+BOOST_AUTO_TEST_CASE(CartPoleWithDoubleDQN)
+{
+  // Set up the network.
+  FFN<MeanSquaredError<>, GaussianInitialization> model;
+  model.Add<Linear<>>(4, 128);
+  model.Add<ReLULayer<>>();
+  model.Add<Linear<>>(128, 128);
+  model.Add<ReLULayer<>>();
+  model.Add<Linear<>>(128, 2);
+
+  // Set up the optimizer.
+  StandardSGD<decltype(model)> opt(model, 0.0001, 2);
+
+  // Set up the policy and replay method.
+  GreedyPolicy<CartPole> policy(1.0, 1000, 0.1);
+  RandomReplay<CartPole> replayMethod(10, 10000);
+
+  // Set up the DQN agent.
+  QLearning<CartPole, decltype(model), decltype(opt), decltype(policy)>
+      agent(model, opt, 0.9, policy, replayMethod, 100, 100, true, 200);
+
+  arma::running_stat<double> averageReturn;
+  while (true)
+  {
```

Comment:
Should we set a maximum number of iterations here?  I think that if there is a bug, this test might not terminate.